### ISSUE TITLE
Pytorch optimizer fails when initialized in workflow

### ISSUE THREAD

**Issue by ineporozhnii** [AgnostiqHQ/covalent#1418]
### Environment

- **Covalent version**: 0.202.0.post1
- **Python version**: Python 3.8.9
- **Operating system**: macOS 12.6 (Monterey)

### What is happening?

If Pytorch optimizer (torch.optim.SGD) is created in the workflow function and passed to the training function, the model does not train (loss doesn’t decrease).  

### How can we reproduce the issue?

To reproduce the problem we can use code from the Covalent MNIST tutorial (https://covalent.readthedocs.io/en/latest/tutorials/0_ClassicalMachineLearning/mnist_images/source.html). If instead of initializing optimizer in train_model function as it is done in the tutorial we create optimizer in workflow and pass it to the train_model function, the model fails to train (loss does not decrease as shown in the output).
```
def workflow(
    model: NeuralNetwork,
    epochs: int = 2,
    batch_size_train: int = 64,
    batch_size_test: int = 1000,
    learning_rate: float = 0.01,
    momentum: float = 0.5,
    log_interval: int = 200,
    loss_fn: callable = F.nll_loss,
):
    """MNIST classifier training workflow"""

    train_dataloader = data_loader(batch_size=batch_size_train, train=True)
    test_dataloader = data_loader(batch_size=batch_size_test, train=False)
    # Create optimizer in workflow
    optimizer = torch.optim.SGD(model.parameters(), learning_rate, momentum)
    train_losses, train_counter, test_losses = [], [], []
    model, optimizer = train_model(
        # Pass optimizer to train_model function
        optimizer = optimizer, 
        train_dataloader=train_dataloader,
        train_losses=train_losses,
        train_counter=train_counter,
        log_interval=log_interval,
        model=model,
        learning_rate=learning_rate,
        momentum=momentum,
        loss_fn=loss_fn,
        epochs=epochs
    )
    test(dataloader=test_dataloader, model=model, loss_fn=loss_fn
    return train_losses, train_counter, test_losses
```
Output: 
```
Epoch 1
-------------------------------
loss: 2.314146  [    0/60000]
loss: 2.304057  [12800/60000]
loss: 2.306395  [25600/60000]
loss: 2.317843  [38400/60000]
loss: 2.303548  [51200/60000]
Epoch 2
-------------------------------
loss: 2.318636  [    0/60000]
loss: 2.287628  [12800/60000]
loss: 2.316385  [25600/60000]
loss: 2.292093  [38400/60000]
loss: 2.306303  [51200/60000]
```

### What should happen?

The loss should decrease with each iteration (weights update) and the output should look like this:
```
Epoch 1
-------------------------------
loss: 2.323291  [    0/60000]
loss: 2.232677  [12800/60000]
loss: 1.339883  [25600/60000]
loss: 1.063978  [38400/60000]
loss: 0.750546  [51200/60000]
Epoch 2
-------------------------------
loss: 0.631298  [    0/60000]
loss: 0.683952  [12800/60000]
loss: 0.445413  [25600/60000]
loss: 0.370649  [38400/60000]
loss: 0.395632  [51200/60000]
```
This output can be achieved by creating an optimizer in the train_model function instead of the workflow.

**Comment by cjao** [comment]
Hello @ineporozhnii! Thanks for trying out the tutorial. You have just stumbled on a rather subtle characteristic of heterogeneous workflows.

Remember that different electrons can in general run on different computers, typically at different times. Thus, object references between different electrons have no meaning when a workflow is dispatched to Covalent. 

More concretely, suppose `object_1` is instantiated in `electron_1` and `electron_2` contains a reference `object_1_ref`. All is well when the workflow runs as a normal Python function because the objects belong to the same process and therefore share the same memory space; `object_1_ref` is simply the memory address of `object_1`. Methods that mutate the underlying object, such as `object_1_ref.update_in_place()`, work as expected.

But suppose now that `electron_2` runs in a different process or even on a different computer. Since the electrons are now segregated from each other in memory, the memory address `object_1_ref` in `electron_2` is now meaningless. `electron_1`'s computer might have already shut down by the time `electron_2` runs.

The rule of thumb to keep in mind is that Covalent makes **deep copies** of all arguments passed to electrons. Now let's look at the your modified workflow:

```python

    # Create optimizer in workflow
    optimizer = torch.optim.SGD(model.parameters(), learning_rate, momentum)
    train_losses, train_counter, test_losses = [], [], []
    model, optimizer = train_model(
        # Pass optimizer to train_model function
        optimizer = optimizer, 
        train_dataloader=train_dataloader,
        train_losses=train_losses,
        train_counter=train_counter,
        log_interval=log_interval,
        model=model,
        learning_rate=learning_rate,
        momentum=momentum,
        loss_fn=loss_fn,
        epochs=epochs
    )
```

Here, `optimizer` holds references to the parameters of the `model` passed as a workflow input. But now inside the `train_model` electron
```python
def train_model(
    train_dataloader: torch.utils.data.dataloader.DataLoader,
    train_losses: list,
    train_counter: int,
    log_interval: int,
    model: NeuralNetwork,
    learning_rate: float,
    momentum: float,
    loss_fn: callable,
    epochs: int,
    results_dir: str = "~/data/mnist/results/",
) -> Tuple[NeuralNetwork,]:
    """Train neural network model."""

    results_dir = Path(results_dir).expanduser()
    results_dir.mkdir(parents=True, exist_ok=True)

    for epoch in range(1, epochs + 1):
        print(f"Epoch {epoch}\n-------------------------------")
        model, optimizer = train_over_one_epoch(
            dataloader=train_dataloader,
            model=model,
            optimizer=optimizer,
            train_losses=train_losses,
            train_counter=train_counter,
            log_interval=log_interval,
            epoch=epoch,
            loss_fn=loss_fn,
        )

    # Save model and optimizer
    torch.save(model.state_dict(), f"{results_dir}model.pth")
    torch.save(optimizer.state_dict(), f"{results_dir}optimizer.pth")
    return model, optimizer
```
`optimizer` is a **deep copy** of the optimizer passed to the electron, and similarly for `model`. The relation between `optimizer` and `model` is broken.

This is why the tutorial instantiates the `optimizer` inside the `train_model` electron; because all code within an electron executes in the same process, object references work as expected.

**Comment by ineporozhnii** [comment]
@cjao Thank you very much for the detailed reply and explanation! One more thing I wanted to mention is that MNIST tutorial seems to have an incorrect output displayed (likely caused by the issue described above). 
<img width="1258" alt="image" src="https://user-images.githubusercontent.com/71763551/202294565-f0b1b6ff-106b-45a7-8462-8573ec8e077d.png">

**Comment by cjao** [comment]
Thanks for pointing that out. That image needs to be fixed.
